comRamona/assignment2/Question1.java

```diff
@@ -0,0 +1,91 @@
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Given a binary tree and a key, write a function that prints all the ancestors of the key in the given
+ * binary tree.
+ */
+public class Question1 {
+    /**
+     * Helper method to find the path to a node and append nodes on the path to the resultPath list(mutable)
+     *
+     * @param node       current root of search
+     * @param target     target value
+     * @param resultPath list to append path
+     * @param <T>        data type of the nodes
+     * @return
+     */
+
+    private static <T> MyNode findAncestorsHelper(MyNode<T> node, MyNode<T> target, List<T> resultPath) {
+        if (node == null) {
+            return null;
+        }
+        if (node.getData().equals(target.getData())) {
+            return node;
+        }
+        MyNode leftSearch = findAncestorsHelper(node.getLeft(), target, resultPath);
+        if (leftSearch != null) {
+            resultPath.add(node.getData());
+            return node;
+        }
+        MyNode rightSearch = findAncestorsHelper(node.getRight(), target, resultPath);
+        if (rightSearch != null) {
+            resultPath.add(node.getData());
+            return node;
+        }
+        return null;
+    }
+
+    /**
+     * @param root   Root node of the tree
+     * @param target Target value to find
+     * @param <T>    data type of the nodes
+     * @return list of ancestors of the target node, null if node not found
+     */
+    public static <T> List<T> findAncestors(MyNode<T> root, MyNode<T> target) {
+        List<T> resultPath = new LinkedList<>();
+        MyNode result = findAncestorsHelper(root, target, resultPath);
+        if(result == null){
+            return null;
+        }
+        return resultPath;
+    }
+
+    public static void main(String[] args) {
+        /*
+         *       a
+         *      /  \
+         *     b    c
+         *   / \     \
+         *  d  e      f
+         * / \
+         * g  h
+         *
+         */
+        MyNode<Character> a = new MyNode<>('a');
```

Comment:
How could we make this cleaner?
What if you created the root of the tree and then called:
root.addChild('b')
root.addChild('c')
root.findInChildren('b').addChild('d')
and so on. What would be the advantages or disadvantages?